### PR TITLE
Implement async loading to unblock datasource update request from HTTP

### PR DESCRIPTION
**Pull request by tribal-tec**

### CODE REVIEW

File: brayns/io/algorithms/MetaballsGenerator.h [hunk 1]
```diff
@@ -44,6 +44,7 @@ class MetaballsGenerator
      * @param threshold Points in 3D space that fall below the threshold
      *        (when run through the function) are ONE, while points above the
      *        threshold are ZERO
+     * @param materials
```
Comment (favreau) [inline brayns/io/algorithms/MetaballsGenerator.h]: Missing doc?

Comment (tribal-tec) [reply brayns/io/algorithms/MetaballsGenerator.h]: yeah, tell me :) you missed the material completely

File: brayns/Brayns.cpp [hunk 2]
```diff
@@ -115,49 +123,90 @@ struct Brayns::Impl
         if (!_engine)
             throw std::runtime_error("Unsupported engine: " + engineName);
 
+        _engine->getFrameBuffer().clear();
```
Comment (rdumusc) [inline brayns/Brayns.cpp]: the frameBuffer isn't cleared initially?

File: brayns/Brayns.cpp [hunk 3]
```diff
@@ -115,49 +123,90 @@ struct Brayns::Impl
...
+        _engine->setReady(false);
+
+        // if engine does not support scene reloading, recreate it first
+        if (!_engine->getScene().supportsUnloading())
+            createEngine();
+
+        _engine->setLastOperation("");
+        _engine->setLastProgress(0);
+
+        auto loadFunc = [&] {
```
Comment (rdumusc) [inline brayns/Brayns.cpp]: make this a _loadScene member function (or free function?), buildScene() is already too long

File: brayns/Brayns.cpp [hunk 4]
```diff
@@ -115,49 +123,90 @@ struct Brayns::Impl
...
+            Progress loadingProgress(
+                "Loading scene...",
+                LOADING_PROGRESS_DATA + 3 * LOADING_PROGRESS_STEP,
+                [&](const std::string& msg, const size_t current,
```
Comment (rdumusc) [inline brayns/Brayns.cpp]: [this], in general don't capture more than is required

File: brayns/Brayns.cpp [hunk 5]
```diff
@@ -283,14 +336,30 @@ struct Brayns::Impl
...
 
 #if (BRAYNS_USE_DEFLECT || BRAYNS_USE_NETWORKING)
         _executePlugins(windowSize);
 #endif
+
+        if (_dataLoadingFuture.valid())
+        {
+#ifdef BRAYNS_USE_LUNCHBOX
+            if (!_parametersManager->getApplicationParameters()
+                     .getSynchronousMode() &&
```
Comment (rdumusc) [inline brayns/Brayns.cpp]: it's the second occurrence of this call which does not fit on a single line, a helper function would be nicer (_useSynchronousMode())

File: brayns/Brayns.cpp [hunk 6]
```diff
@@ -321,15 +392,31 @@ struct Brayns::Impl
...
+    void _loadData(Progress& loadingProgress)
     {
         auto& geometryParameters = _parametersManager->getGeometryParameters();
         auto& volumeParameters = _parametersManager->getVolumeParameters();
         auto& sceneParameters = _parametersManager->getSceneParameters();
         auto& scene = _engine->getScene();
 
+        size_t nextTic = 0;
+        const size_t tic = LOADING_PROGRESS_DATA;
+        auto progressUpdate = [&](const std::string& msg, const size_t current,
```
Comment (rdumusc) [inline brayns/Brayns.cpp]: this is a function, rename to updateProgress() ?

File: brayns/Brayns.cpp [hunk 7]
```diff
@@ -407,18 +507,21 @@ struct Brayns::Impl
     /**
         Loads data from a PDB file (command line parameter --pdb-file)
     */
-    void _loadPDBFolder()
+    void _loadPDBFolder(const Progress::ProgressUpdateCallback& progressUpdate)
```
Comment (rdumusc) [inline brayns/Brayns.cpp]: Progress::ProgressUpdateCallback is redundant, just Progress::UpdateCallback

File: brayns/Brayns.cpp [hunk 8]
```diff
@@ -450,14 +553,15 @@ struct Brayns::Impl
...
-    void _loadXYZBFile()
+    void _loadXYZBFile(const Progress::ProgressUpdateCallback& progressUpdate)
     {
         // Load XYZB File
         auto& geometryParameters = _parametersManager->getGeometryParameters();
         auto& scene = _engine->getScene();
         BRAYNS_INFO << "Loading XYZB file " << geometryParameters.getXYZBFile()
                     << std::endl;
         XYZBLoader xyzbLoader(geometryParameters);
+        xyzbLoader.setProgressUpdate(progressUpdate);
```
Comment (rdumusc) [inline brayns/Brayns.cpp]: setProgressCallback

File: brayns/common/Progress.h [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+ */
+class Progress
+{
+public:
+    /**
+     * The callback for each progress update with the signature (message,
+     * current count, expected count)
+     */
+    using ProgressUpdateCallback =
+        std::function<void(const std::string&, size_t, size_t)>;
```
Comment (rdumusc) [inline brayns/common/Progress.h]: is it necessary to pass the expected count at every step? or could we use a base 100 or a float?

File: brayns/common/Progress.h [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+    /**
+     * Setup the progress and already report progress '0' on the given callback.
+     * @param message the current operation that takes place
+     * @param expectedCount the expected count of updates when the reporting
+     *        should finish
+     * @param updateCallback a user-defined callback to be called on each
+     *        update.
+     */
+    Progress(const std::string& message, unsigned long expectedCount,
```
Comment (rdumusc) [inline brayns/common/Progress.h]: unsigned long ->size_t for consistency

File: brayns/common/engine/Engine.h [hunk 11]
```diff
@@ -145,12 +151,25 @@ class Engine
...
      * @return the last operation processed by the engine
      */
-    const std::string getLastOperation() const { return _lastOperation; }
+    const std::string& getLastOperation() const { return _lastOperation; }
     /**
      * @return the last normalized progress value (0..1) emitted by the engine
      */
     float getLastProgress() const { return _lastProgress; }
     /**
+     * Set the last normalized progress value (0..1) for any current operation.
```
Comment (rdumusc) [inline brayns/common/engine/Engine.h]: swap doxygen string with next function

File: brayns/io/ProgressReporter.h [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+{
+/**
+ * A base class for any loader who wishes to report progress during loading
+ * operations.
+ */
+class ProgressReporter
+{
+public:
+    ProgressReporter() = default;
+    ~ProgressReporter() = default;
```
Comment (rdumusc) [inline brayns/io/ProgressReporter.h]: virtual?!
default constructor needed explictly?

File: brayns/Brayns.cpp (hunk 2 above)
Comment (tribal-tec) [reply brayns/Brayns.cpp]: not explicitly in Brayns, ospray and optix might do it.

**Pull request merged**